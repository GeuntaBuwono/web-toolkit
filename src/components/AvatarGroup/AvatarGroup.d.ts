type AvatarItem = Required<Omit<AvatarType, 'size'>>;

type AvatarGroup = {
	maxLength: number;
	avatars: Array<AvatarItem>;
	size?: AvatarSize;
};
