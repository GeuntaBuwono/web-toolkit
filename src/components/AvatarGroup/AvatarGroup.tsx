import Avatar, {AvatarProps, AvatarSize} from '@components/Avatar/Avatar';

export type AvatarItem = Required<Omit<AvatarProps, 'size'>>;

export type AvatarGroupProps = {
	maxLength: number;
	avatars: Array<AvatarItem>;
	size?: AvatarSize;
};

const AVATAR_GARP_STACK_SIZE = 'ml-[-0.75rem]';

const AvatarGroup = ({size = 'md', maxLength, avatars}: AvatarGroupProps) => {
	if (maxLength > 0) {
		return (
			<div className="flex flex-row">
				{avatars.map((avatar, index) => {
					if (index < maxLength) {
						return (
							<div className={`${AVATAR_GARP_STACK_SIZE}`} key={index}>
								<Avatar {...avatar} size={size} name="asd" />
							</div>
						);
					}
					return null;
				})}
				{avatars.length > maxLength && (
					<div className={`${AVATAR_GARP_STACK_SIZE}`}>
						<Avatar name={`+${avatars.length - maxLength}`} size={size} />
					</div>
				)}
			</div>
		);
	}
	return null;
};

export default AvatarGroup;
