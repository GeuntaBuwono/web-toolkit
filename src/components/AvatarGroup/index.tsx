import Avatar from '@components/Avatar';

const AVATAR_GARP_STACK_SIZE = 'ml-[-0.75rem]';

const AvatarGroup = ({size = 'md', maxLength, avatars}: AvatarGroup) => {
	if (maxLength > 0) {
		return (
			<div className="flex flex-row">
				{avatars.map((avatar, index) => {
					if (index < maxLength) {
						return (
							<div className={`${AVATAR_GARP_STACK_SIZE}`} key={index}>
								<Avatar {...avatar} size={size} />
							</div>
						);
					}
					return null;
				})}
				{avatars.length > maxLength && (
					<div className={`${AVATAR_GARP_STACK_SIZE}`}>
						<Avatar name={`+${avatars.length - maxLength}`} size={size} />
					</div>
				)}
			</div>
		);
	}
	return null;
};

export default AvatarGroup;
