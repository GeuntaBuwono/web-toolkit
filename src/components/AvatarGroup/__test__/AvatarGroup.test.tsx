import {render} from '@testing-library/react';
import AvatarGroup from '@components/AvatarGroup';

const testAvatarGroupProps: AvatarGroup = {
	maxLength: 3,
	avatars: [
		{
			name: 'budi',
			src: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/160.jpg',
			alt: 'expectedAvatarAlt',
		},
		{
			name: 'budi tomi',
			src: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/161.jpg',
			alt: 'expectedAvatarAlt',
		},
		{
			name: 'budi abi',
			src: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/162.jpg',
			alt: 'expectedAvatarAlt',
		},
		{
			name: 'budi herman',
			src: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/164.jpg',
			alt: 'expectedAvatarAlt',
		},
		{
			name: 'budi budi',
			src: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/165.jpg',
			alt: 'expectedAvatarAlt',
		},
		{
			name: 'budi imron',
			src: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/166.jpg',
			alt: 'expectedAvatarAlt',
		},
		{
			name: 'budi jaya',
			src: 'https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/167.jpg',
			alt: 'expectedAvatarAlt',
		},
	],
};

describe('AvatarGroup', () => {
	test('renders default', () => {
		const {getByText} = render(<AvatarGroup {...testAvatarGroupProps} />);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(getByText('+4')).toBeTruthy();
		expect(testAvatarElement.className).toContain('h-10 w-10');
	});

	test('renders with size', () => {
		const {getByText} = render(
			<AvatarGroup size="sm" {...testAvatarGroupProps} />,
		);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(getByText('+4')).toBeTruthy();
		expect(testAvatarElement.className).toContain('h-8 w-8');
	});

	test('renders with maxLength 4', () => {
		const {getByText} = render(
			<AvatarGroup {...testAvatarGroupProps} size="sm" maxLength={4} />,
		);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(getByText('+3')).toBeTruthy();
		expect(testAvatarElement.className).toContain('h-8 w-8');
	});

	test('renders with maxLength MORE or EQUAL with avatars length', () => {
		render(<AvatarGroup {...testAvatarGroupProps} size="sm" maxLength={10} />);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;
		const textAvatarGroupMoreElement = document.querySelector(
			'p',
		) as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(textAvatarGroupMoreElement).toBeNull();
		expect(testAvatarElement.className).toContain('h-8 w-8');
	});

	test('renders with maxLength 0', () => {
		render(<AvatarGroup {...testAvatarGroupProps} size="sm" maxLength={0} />);
		const testAvatarElement = document.querySelector('img') as HTMLImageElement;
		const textAvatarGroupMoreElement = document.querySelector(
			'p',
		) as HTMLImageElement;

		expect(testAvatarElement).toMatchSnapshot();
		expect(testAvatarElement).toBeNull();
		expect(textAvatarGroupMoreElement).toBeNull();
	});
});
