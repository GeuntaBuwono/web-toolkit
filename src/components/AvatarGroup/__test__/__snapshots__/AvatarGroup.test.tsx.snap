// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AvatarGroup renders default 1`] = `
<img
  alt="expectedAvatarAlt"
  class="inline-block rounded-full ring-2 ring-white h-10 w-10"
  src="https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/160.jpg"
/>
`;

exports[`AvatarGroup renders with maxLength 0 1`] = `null`;

exports[`AvatarGroup renders with maxLength 4 1`] = `
<img
  alt="expectedAvatarAlt"
  class="inline-block rounded-full ring-2 ring-white h-8 w-8"
  src="https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/160.jpg"
/>
`;

exports[`AvatarGroup renders with maxLength MORE or EQUAL with avatars length 1`] = `
<img
  alt="expectedAvatarAlt"
  class="inline-block rounded-full ring-2 ring-white h-8 w-8"
  src="https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/160.jpg"
/>
`;

exports[`AvatarGroup renders with size 1`] = `
<img
  alt="expectedAvatarAlt"
  class="inline-block rounded-full ring-2 ring-white h-8 w-8"
  src="https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/160.jpg"
/>
`;
